### PR TITLE
Machine based reading order integration: some refactoring and fixes

### PR DESCRIPTION
**Pull request by bertsky**
I tried to run the new branch on https://github.com/OCR-D/gt_structure_all/tree/main/datasets, but ran into a couple of problems:

1. CUDA libraries (specifically, libcudnn) could not be installed properly, because the new OCR feature depends on Pytorch, which explicitly depends on (and is dynamically linked against) a newer version of nvidia-cudnn than the one Tensorflow implicitly needs (and is dynamically loaded)
   - [x] fixed by manually downgrading, but to avoid that problem for unsuspecting users, I also made the OCR feature (and its dependencies) into an optional feature; same goes for matplotlib, which could drag in X11 libs IIRC
2. CUDA OOM with `dir_in` mode after a few hundred pages
   - [x] fixed by ensuring no models are reloaded in that mode, and adding some `gc.collect`
   - [x] probably also be removed an explicit `del` (don't recall if this made a difference)
3. no log output
   - [x] fixed by setting the log level for our actual logger `eynollah` instead of `ocrd_utils.setOverrideLogLevel` (which only affects `ocrd.*` and some preconfigured loggers)
4. no deskewing (always 0° results)
   - [x] fixed by correct indentation for aggregating results (was behind exception handler)
5. non-termination (sleep state) after a few hundred pages
   - [x] fixed by using multiprocessing.Pool instead of custom Process/Queue loops for deskewing and contour extraction
6. segfault after a few hundred pages
   - [x] fixed by avoiding loop body (leading towards `cv2.resize` on zero-channel label array) if no text regions detected

In doing so, or rather in order to achieve that, I had to simplify and refactor here and there to make it readable (to me). There were lots of extremely long lines, code duplication, unnecessary indentation etc.

In particular, I rewrote the parallel subprocessing by utilising concurrent.futures.ProcessPoolExecutor, and maximally reusing the executor instance to avoid the overhead of setting up processes, queues and threads. In my measurements, this reduced the average runtime per page from 26.8 secs to 14.3 secs. GPU utilization is still peaky, though:

eynollah-light-cl-ocrd-gtsa-pool

(This interval was taken over 9 min or a few dozen pages.) I will address CPU-GPU pipelining another time.

I also added the detected deskewing angle to the regions as `@orientation` attribute. 

Moreover, I introduced `--overwrite` to ignore existing output XMLs, and changed the default behaviour to _skip_ them (so one can easily complete a directory if a previous run failed or new images were added).

@vahidrezanezhad, since there are so many large, but rather cosmetic diffs, I recommend going through the changes commit by commit instead of the aggregated file by file view.

### CODE REVIEW

File: src/eynollah/utils/separate_lines.py [hunk 1]
```diff
@@ -1845,7 +1845,7 @@ def do_work_of_slopes_new_light(
     mask_only_con_region = np.zeros(textline_mask_tot_ea.shape)
     mask_only_con_region = cv2.fillPoly(mask_only_con_region, pts=[contour_par], color=(1, 1, 1))
 
-    if self.textline_light:
+    if textline_light:
```
Comment (bertsky) [inline src/eynollah/utils/separate_lines.py:1848]: Oops, sorry, should have seen that!

File: src/eynollah/utils/separate_lines.py [hunk 2]
```diff
@@ -1857,4 +1857,4 @@ def do_work_of_slopes_new_light(
         all_text_region_raw[mask_only_con_region == 0] = 0
         cnt_clean_rot = textline_contours_postprocessing(all_text_region_raw, slope_deskew, contour_par, box_text)
 
-    return cnt_clean_rot, box_text, contour, contour_par, crop_coor, index_r_con, slope
+    return cnt_clean_rot, box_text, contour, contour_par, crop_coor, index_r_con, slope_deskew
```
Comment (bertsky) [inline src/eynollah/utils/separate_lines.py:1860]: ditto

File: src/eynollah/eynollah.py [hunk 3]
```diff
@@ -4487,12 +4487,9 @@ def run(self):
                     continue
                 else:
                     return pcgts
-
-            img_res, is_image_enhanced, num_col_classifier, num_column_is_classified = self.run_enhancement(self.light_version)
-            self.logger.info("Enhancing took %.1fs ", time.time() - t0)
-            #print("text region early -1 in %.1fs", time.time() - t0)
-            t1 = time.time()
```
Comment (bertsky) [inline src/eynollah/eynollah.py]: IMO it does not make sense to move that from the lower indentation level to all the conditional branches in a copycat fashion. `img_res` seems to be needed everywhere below, so why not just compute it once here?

(See follow-up comments below.)

File: src/eynollah/eynollah.py [hunk 4]
```diff
@@ -4522,467 +4519,471 @@ def run(self):
                 polygons_lines_xml = []
                 contours_tables = []
                 ocr_all_textlines = None
-
                 pcgts = self.writer.build_pagexml_no_full_layout(cont_page, page_coord, order_text_new, id_of_texts_tot,
                                                                  all_found_textline_polygons, page_coord, polygons_of_images, polygons_of_marginals,
                                                                  all_found_textline_polygons_marginals, all_box_coord_marginals, slopes, slopes_marginals,
                                                                  cont_page, polygons_lines_xml, contours_tables, ocr_all_textlines)
                 if self.dir_in:
+                    self.writer.write_pagexml(pcgts)
```
Comment (bertsky) [inline src/eynollah/eynollah.py:4631]: Sry, I missed that one.

File: src/eynollah/eynollah.py [hunk 5]
```diff
@@ -4522,467 +4519,471 @@ def run(self):
...
                     continue
                 else:
                     return pcgts
+            if not self.extract_only_images and not self.skip_layout_and_reading_order:
```
Comment (bertsky) [inline src/eynollah/eynollah.py:4635]: This condition is completely redundant: both `extract_only_images` and `skip_layout_and_reading_order` had their condition checked and the function return by now, so you are essentially saying `if True` here.

Comment (bertsky) [reply src/eynollah/eynollah.py:4635]: The huge diff in all remaining hunks (which could be much smaller if `diff --minimal` were used) is purely indentation AFAICS, correct?

(Okay, except for the two additional return tuple elements of `filter_contours_without_textline_inside`.)

Comment (vahidrezanezhad) [reply src/eynollah/eynollah.py:4635]: > The huge diff in all remaining hunks (which could be much smaller if `diff --minimal` were used) is purely indentation AFAICS, correct?
> 
> (Okay, except for the two additional return tuple elements of `filter_contours_without_textline_inside`.)

True. The only key things that were missed were extracting_only_images and skip_layout_ro.

Comment (bertsky) [reply src/eynollah/eynollah.py]: So I removed the redundant conditional and reduced the indentation of the remaining body of the function again.

File: src/eynollah/eynollah.py (hunk 3 above)
Comment (bertsky) [reply src/eynollah/eynollah.py]: I have moved it up again, so that code does not need to be repeated.

**Comment by bertsky** [comment]
I dug through most of the code base, still watching out for places that might act like a memory leak. To be able to read and understand the code, I had to do further simplification and styling (esp. wrapping overlong lines) – I hope you don't mind these changes. I have tested them in various modes, no differences in output so far.

Note: in [0ae28f7](https://github.com/qurator-spk/eynollah/pull/142/commits/0ae28f7d3ef33d6bf3650b99bc7646c3234341d1) I switched from stdlib ProcessPoolExecutor to [loky](https://github.com/joblib/loky), which is the origin of many bugfixes in stdlib between Python 3.9 and 3.11, which in my experience are needed for robustness, but won't be backported for 3.8. (When in the distant future we will have moved to 3.11 anyway, we can remove that dependency and switch back to stdlib.)

I am still hunting the OOM failures (via instrumentation and monitoring), so stay tuned. (I'll judiciously compare performance gains/losses of the recent changes as soon as the code is sufficiently stable.)
